File: typescript/packages/mechanisms/evm/src/exact/facilitator/eip3009-utils.ts
import { PaymentRequirements, VerifyResponse } from "@x402/core/types";
import { encodeFunctionData, getAddress, Hex, parseErc6492Signature, parseSignature } from "viem";
import { eip3009ABI } from "../../constants";
import { multicall, ContractCall, RawContractCall } from "../../multicall";
import { FacilitatorEvmSigner } from "../../signer";
import { ExactEIP3009Payload } from "../../types";
import * as Errors from "./errors";

export interface Eip6492Deployment {
  factoryAddress: `0x${string}`;
  factoryCalldata: `0x${string}`;
}

/**
 * Outcome of a transfer simulation. `error` is populated only when the underlying `eth_call`
 * threw (revert or transport), letting callers distinguish a contract revert from an RPC failure.
 */
export interface SimulateEip3009Result {
  ok: boolean;
  error?: unknown;
}

/**
 * Simulates transferWithAuthorization via eth_call.
 * Returns true if simulation succeeded, false if it failed.
 *
 * @param signer - EVM signer for contract reads
 * @param erc20Address - ERC-20 token contract address
 * @param payload - EIP-3009 transfer authorization payload
 * @param eip6492Deployment - Optional EIP-6492 factory info for undeployed smart wallets
 *
 * @returns true if simulation succeeded, false if it failed
 */
export async function simulateEip3009Transfer(
  signer: FacilitatorEvmSigner,
  erc20Address: `0x${string}`,
  payload: ExactEIP3009Payload,
  eip6492Deployment?: Eip6492Deployment,
): Promise<boolean> {
  return (await simulateEip3009TransferResult(signer, erc20Address, payload, eip6492Deployment)).ok;
}

/**
 * Like {@link simulateEip3009Transfer} but returns the thrown error (if any) alongside the
 * boolean outcome, so callers can tell a contract revert apart from a transport/RPC failure.
 *
 * @param signer - EVM signer for contract reads
 * @param erc20Address - ERC-20 token contract address
 * @param payload - EIP-3009 transfer authorization payload
 * @param eip6492Deployment - Optional EIP-6492 factory info for undeployed smart wallets
 * @returns The simulation outcome and, on failure via a thrown error, that error.
 */
export async function simulateEip3009TransferResult(
  signer: FacilitatorEvmSigner,
  erc20Address: `0x${string}`,
  payload: ExactEIP3009Payload,
  eip6492Deployment?: Eip6492Deployment,
): Promise<SimulateEip3009Result> {
  const auth = payload.authorization;
  const transferArgs = [
    getAddress(auth.from),
    getAddress(auth.to),
    BigInt(auth.value),
    BigInt(auth.validAfter),
    BigInt(auth.validBefore),
    auth.nonce,
  ] as const;

  if (eip6492Deployment) {
    const { signature: innerSignature } = parseErc6492Signature(payload.signature!);
    const transferCalldata = encodeFunctionData({
      abi: eip3009ABI,
      functionName: "transferWithAuthorization",
      args: [...transferArgs, innerSignature],
    });

    try {
      const results = await multicall(signer.readContract.bind(signer), [
        {
          address: getAddress(eip6492Deployment.factoryAddress),
          callData: eip6492Deployment.factoryCalldata,
        } satisfies RawContractCall,
        {
          address: erc20Address,
          callData: transferCalldata,
        } satisfies RawContractCall,
      ]);

      return { ok: results[1]?.status === "success" };
    } catch (error) {
      return { ok: false, error };
    }
  }

  const sig = payload.signature!;
  const sigLength = sig.startsWith("0x") ? sig.length - 2 : sig.length;
  const isECDSA = sigLength === 130;

  try {
    if (isECDSA) {
      const parsedSig = parseSignature(sig);
      await signer.readContract({
        address: erc20Address,
        abi: eip3009ABI,
        functionName: "transferWithAuthorization",
        args: [
          ...transferArgs,
          (parsedSig.v as number | undefined) ?? parsedSig.yParity,
          parsedSig.r,
          parsedSig.s,
        ],
      });
    } else {
      await signer.readContract({
        address: erc20Address,
        abi: eip3009ABI,
        functionName: "transferWithAuthorization",
        args: [...transferArgs, sig],
      });
    }
    return { ok: true };
  } catch (error) {
    return { ok: false, error };
  }
}

/**
 * After simulation fails, runs a single diagnostic multicall to determine the most specific error reason.
 * Checks balanceOf, name, version and authorizationState in one RPC round-trip.
 *
 * @param signer - EVM signer used for the payment
 * @param erc20Address - Address of the ERC-20 token contract
 * @param payload - The EIP-3009 transfer authorization payload
 * @param requirements - Payment requirements to validate against
 * @param amountRequired - Required amount for the payment (balance check)
 *
 * @returns Promise resolving to the verification result with validity and optional invalid reason
 */
export async function diagnoseEip3009SimulationFailure(
  signer: FacilitatorEvmSigner,
  erc20Address: `0x${string}`,
  payload: ExactEIP3009Payload,
  requirements: PaymentRequirements,
  amountRequired: string,
): Promise<VerifyResponse> {
  const payer = payload.authorization.from;

  const diagnosticCalls: ContractCall[] = [
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "balanceOf",
      args: [payload.authorization.from],
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "name",
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "version",
    },
    {
      address: erc20Address,
      abi: eip3009ABI,
      functionName: "authorizationState",
      args: [payload.authorization.from, payload.authorization.nonce],
    },
  ];

  try {
    const results = await multicall(signer.readContract.bind(signer), diagnosticCalls);

    const [balanceResult, nameResult, versionResult, authStateResult] = results;

    if (authStateResult.status === "failure") {
      return { isValid: false, invalidReason: Errors.ErrEip3009NotSupported, payer };
    }

    if (authStateResult.status === "success" && authStateResult.result === true) {
      return { isValid: false, invalidReason: Errors.ErrEip3009NonceAlreadyUsed, payer };
    }

    if (
      nameResult.status === "success" &&
      requirements.extra?.name &&
      nameResult.result !== requirements.extra.name
    ) {
      return { isValid: false, invalidReason: Errors.ErrEip3009TokenNameMismatch, payer };
    }

    if (
      versionResult.status === "success" &&
      requirements.extra?.version &&
      versionResult.result !== requirements.extra.version
    ) {
      return { isValid: false, invalidReason: Errors.ErrEip3009TokenVersionMismatch, payer };
    }

    if (balanceResult.status === "success") {
      const balance = balanceResult.result as bigint;
      if (balance < BigInt(amountRequired)) {
        return {
          isValid: false,
          invalidReason: Errors.ErrEip3009InsufficientBalance,
          payer,
        };
      }
    }
  } catch {
    // Diagnostic multicall failed — fall through to generic error
  }

  return { isValid: false, invalidReason: Errors.ErrEip3009SimulationFailed, payer };
}

/**
 * Maps an EIP-3009 contract revert error to a specific error code.
 * Falls back to ErrTransactionFailed when the revert reason is unknown.
 *
 * @param error - The error thrown during transfer execution
 * @returns A specific error reason string
 */
export function parseEip3009TransferError(error: unknown): string {
  const msg = error instanceof Error ? error.message : String(error);
  if (/authorization.*(expired|valid before)/i.test(msg) || /AuthorizationExpired/i.test(msg)) {
    return Errors.ErrValidBeforeExpired;
  }
  if (/authorization.*not.*valid|AuthorizationNotYetValid/i.test(msg)) {
    return Errors.ErrValidAfterInFuture;
  }
  if (/authorization.*used|AuthorizationAlreadyUsed|AuthorizationUsedOrCanceled/i.test(msg)) {
    return Errors.ErrEip3009NonceAlreadyUsed;
  }
  if (/transfer.*exceeds.*balance|insufficient.*balance|ERC20InsufficientBalance/i.test(msg)) {
    return Errors.ErrEip3009InsufficientBalance;
  }
  if (/invalid.*signature|SignerMismatch|InvalidSignatureV|InvalidSignatureS/i.test(msg)) {
    return Errors.ErrInvalidSignature;
  }
  return Errors.ErrTransactionFailed;
}

/**
 * Executes transferWithAuthorization onchain.
 *
 * @param signer - EVM signer for contract writes
 * @param erc20Address - ERC-20 token contract address
 * @param payload - EIP-3009 transfer authorization payload
 * @param dataSuffix - Optional hex bytes to append after the ABI-encoded calldata
 *
 * @returns Transaction hash
 */
export async function executeTransferWithAuthorization(
  signer: FacilitatorEvmSigner,
  erc20Address: `0x${string}`,
  payload: ExactEIP3009Payload,
  dataSuffix?: Hex,
): Promise<Hex> {
  const { signature } = parseErc6492Signature(payload.signature!);
  const signatureLength = signature.startsWith("0x") ? signature.length - 2 : signature.length;
  const isECDSA = signatureLength === 130;

  const auth = payload.authorization;
  const baseArgs = [
    getAddress(auth.from),
    getAddress(auth.to),
    BigInt(auth.value),
    BigInt(auth.validAfter),
    BigInt(auth.validBefore),
    auth.nonce,
  ] as const;

  let signatureArgs: readonly unknown[];
  if (isECDSA) {
    const parsedSig = parseSignature(signature);
    signatureArgs = [
      (parsedSig.v as number | undefined) || parsedSig.yParity,
      parsedSig.r,
      parsedSig.s,
    ];
  } else {
    signatureArgs = [signature];
  }

  return signer.writeContract({
    address: erc20Address,
    abi: eip3009ABI,
    functionName: "transferWithAuthorization",
    args: [...baseArgs, ...signatureArgs],
    dataSuffix,
  });
}

File: python/x402/mechanisms/evm/exact/eip3009_utils.py
"""Shared EIP-3009 helpers for exact EVM facilitators."""

from __future__ import annotations

from dataclasses import dataclass

from ..constants import (
    AUTHORIZATION_STATE_ABI,
    BALANCE_OF_ABI,
    ERR_EIP3009_NOT_SUPPORTED,
    ERR_INSUFFICIENT_BALANCE,
    ERR_NONCE_ALREADY_USED,
    ERR_TOKEN_NAME_MISMATCH,
    ERR_TOKEN_VERSION_MISMATCH,
    ERR_TRANSACTION_SIMULATION_FAILED,
    FUNCTION_TRANSFER_WITH_AUTHORIZATION,
    NAME_ABI,
    TRANSFER_WITH_AUTHORIZATION_BYTES_ABI,
    TRANSFER_WITH_AUTHORIZATION_VRS_ABI,
    VERSION_ABI,
)
from ..eip712 import build_typed_data_for_signing
from ..erc6492 import has_deployment_info, parse_erc6492_signature
from ..multicall import MulticallCall, encode_contract_call, multicall
from ..signer import FacilitatorEvmSigner
from ..types import ERC6492SignatureData, ExactEIP3009Authorization
from ..utils import bytes_to_hex, hex_to_bytes


@dataclass
class ParsedEIP3009Authorization:
    """Parsed authorization values ready for contract calls."""

    from_address: str
    to: str
    value: int
    valid_after: int
    valid_before: int
    nonce: bytes


@dataclass
class EIP3009SignatureClassification:
    """How the facilitator should treat a signature before simulation."""

    valid: bool
    is_smart_wallet: bool
    is_undeployed: bool
    sig_data: ERC6492SignatureData


def parse_eip3009_authorization(
    authorization: ExactEIP3009Authorization,
) -> ParsedEIP3009Authorization:
    """Parse string-encoded authorization fields into contract-call values."""
    nonce = hex_to_bytes(authorization.nonce)
    if len(nonce) != 32:
        raise ValueError(f"invalid nonce length: got {len(nonce)} bytes, want 32")

    return ParsedEIP3009Authorization(
        from_address=authorization.from_address,
        to=authorization.to,
        value=int(authorization.value),
        valid_after=int(authorization.valid_after),
        valid_before=int(authorization.valid_before),
        nonce=nonce,
    )


def classify_eip3009_signature(
    signer: FacilitatorEvmSigner,
    authorization: ExactEIP3009Authorization,
    signature: bytes,
    chain_id: int,
    token_address: str,
    token_name: str,
    token_version: str,
) -> EIP3009SignatureClassification:
    """Classify the signature before deciding whether simulation may rescue it."""
    sig_data = parse_erc6492_signature(signature)
    domain, types, primary_type, message = build_typed_data_for_signing(
        authorization,
        chain_id,
        token_address,
        token_name,
        token_version,
    )

    is_smart_wallet = has_deployment_info(sig_data) or len(sig_data.inner_signature) != 65
    valid = signer.verify_typed_data(
        authorization.from_address,
        domain,
        types,
        primary_type,
        message,
        sig_data.inner_signature,
    )
    if valid:
        return EIP3009SignatureClassification(
            valid=True,
            is_smart_wallet=is_smart_wallet,
            is_undeployed=False,
            sig_data=sig_data,
        )

    code = signer.get_code(authorization.from_address)
    if len(code) > 0:
        return EIP3009SignatureClassification(
            valid=False,
            is_smart_wallet=True,
            is_undeployed=False,
            sig_data=sig_data,
        )

    if has_deployment_info(sig_data):
        return EIP3009SignatureClassification(
            valid=False,
            is_smart_wallet=True,
            is_undeployed=True,
            sig_data=sig_data,
        )

    return EIP3009SignatureClassification(
        valid=False,
        is_smart_wallet=is_smart_wallet,
        is_undeployed=is_smart_wallet,
        sig_data=sig_data,
    )


def simulate_eip3009_transfer(
    signer: FacilitatorEvmSigner,
    token_address: str,
    parsed: ParsedEIP3009Authorization,
    sig_data: ERC6492SignatureData,
) -> bool:
    """Simulate `transferWithAuthorization` and return whether it succeeds."""
    if has_deployment_info(sig_data):
        transfer_calldata = encode_contract_call(
            TRANSFER_WITH_AUTHORIZATION_BYTES_ABI,
            FUNCTION_TRANSFER_WITH_AUTHORIZATION,
            parsed.from_address,
            parsed.to,
            parsed.value,
            parsed.valid_after,
            parsed.valid_before,
            parsed.nonce,
            sig_data.inner_signature,
        )
        try:
            results = multicall(
                signer,
                [
                    MulticallCall(
                        address=bytes_to_hex(sig_data.factory),
                        call_data=sig_data.factory_calldata,
                    ),
                    MulticallCall(address=token_address, call_data=transfer_calldata),
                ],
            )
        except Exception:
            return False
        return len(results) >= 2 and results[1].success

    if len(sig_data.inner_signature) == 65:
        v, r, s = _split_signature_parts(sig_data.inner_signature)
        try:
            signer.read_contract(
                token_address,
                TRANSFER_WITH_AUTHORIZATION_VRS_ABI,
                FUNCTION_TRANSFER_WITH_AUTHORIZATION,
                parsed.from_address,
                parsed.to,
                parsed.value,
                parsed.valid_after,
                parsed.valid_before,
                parsed.nonce,
                v,
                r,
                s,
            )
        except Exception:
            return False
        return True

    try:
        signer.read_contract(
            token_address,
            TRANSFER_WITH_AUTHORIZATION_BYTES_ABI,
            FUNCTION_TRANSFER_WITH_AUTHORIZATION,
            parsed.from_address,
            parsed.to,
            parsed.value,
            parsed.valid_after,
            parsed.valid_before,
            parsed.nonce,
            sig_data.inner_signature,
        )
    except Exception:
        return False
    return True


def diagnose_eip3009_simulation_failure(
    signer: FacilitatorEvmSigner,
    token_address: str,
    authorization: ExactEIP3009Authorization,
    required_amount: int,
    token_name: str,
    token_version: str,
) -> str:
    """Map a failed transfer simulation to the most specific invalid reason."""
    try:
        results = multicall(
            signer,
            [
                MulticallCall(
                    address=token_address,
                    abi=BALANCE_OF_ABI,
                    function_name="balanceOf",
                    args=(authorization.from_address,),
                ),
                MulticallCall(address=token_address, abi=NAME_ABI, function_name="name"),
                MulticallCall(
                    address=token_address,
                    abi=VERSION_ABI,
                    function_name="version",
                ),
                MulticallCall(
                    address=token_address,
                    abi=AUTHORIZATION_STATE_ABI,
                    function_name="authorizationState",
                    args=(authorization.from_address, hex_to_bytes(authorization.nonce)),
                ),
            ],
        )
    except Exception:
        return ERR_TRANSACTION_SIMULATION_FAILED

    if len(results) < 4:
        return ERR_TRANSACTION_SIMULATION_FAILED

    authorization_state = results[3]
    if not authorization_state.success:
        return ERR_EIP3009_NOT_SUPPORTED
    if bool(authorization_state.result):
        return ERR_NONCE_ALREADY_USED

    name_result = results[1]
    if token_name and name_result.success and isinstance(name_result.result, str):
        if name_result.result != token_name:
            return ERR_TOKEN_NAME_MISMATCH

    version_result = results[2]
    if token_version and version_result.success and isinstance(version_result.result, str):
        if version_result.result != token_version:
            return ERR_TOKEN_VERSION_MISMATCH

    balance_result = results[0]
    if balance_result.success:
        try:
            if int(balance_result.result) < required_amount:
                return ERR_INSUFFICIENT_BALANCE
        except (TypeError, ValueError):
            pass

    return ERR_TRANSACTION_SIMULATION_FAILED


def execute_transfer_with_authorization(
    signer: FacilitatorEvmSigner,
    token_address: str,
    parsed: ParsedEIP3009Authorization,
    sig_data: ERC6492SignatureData,
) -> str:
    """Execute `transferWithAuthorization` using the correct ABI overload."""
    if len(sig_data.inner_signature) == 65:
        v, r, s = _split_signature_parts(sig_data.inner_signature)
        return signer.write_contract(
            token_address,
            TRANSFER_WITH_AUTHORIZATION_VRS_ABI,
            FUNCTION_TRANSFER_WITH_AUTHORIZATION,
            parsed.from_address,
            parsed.to,
            parsed.value,
            parsed.valid_after,
            parsed.valid_before,
            parsed.nonce,
            v,
            r,
            s,
        )

    return signer.write_contract(
        token_address,
        TRANSFER_WITH_AUTHORIZATION_BYTES_ABI,
        FUNCTION_TRANSFER_WITH_AUTHORIZATION,
        parsed.from_address,
        parsed.to,
        parsed.value,
        parsed.valid_after,
        parsed.valid_before,
        parsed.nonce,
        sig_data.inner_signature,
    )


def _split_signature_parts(signature: bytes) -> tuple[int, bytes, bytes]:
    if len(signature) != 65:
        raise ValueError(f"invalid ECDSA signature length: expected 65, got {len(signature)}")

    v = signature[64]
    if v in (0, 1):
        v += 27
    return (v, signature[:32], signature[32:64])
